packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+          assert(vrm.isDurable(vref), X`key is not durable`);
+        }
+        if (hasWeakKeys) {
+          vrm.addRecognizableValue(key, entryDeleter);
+        } else {
+          vrm.addReachableVref(vref);
+        }
+      }
+      serializedValue.slots.map(vrm.addReachableVref);
+      syscall.vatstoreSet(keyToDBKey(key), JSON.stringify(serializedValue));
+      ensureEntryCount();
+      entryCount += 1;
+    }
+
+    function set(key, value) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      if (valueSchema) {
+        assert(
+          matches(value, valueSchema),
+          X`invalid value type for collection ${q(label)}`,
+        );
+      }
+      const dbKey = keyToDBKey(key);
+      const rawBefore = syscall.vatstoreGet(dbKey);
+      assert(rawBefore, X`key ${key} not found in collection ${q(label)}`);
+      const before = JSON.parse(rawBefore);
+      const after = serialize(harden(value));
+      if (durable) {
+        after.slots.map(vref =>
+          assert(vrm.isDurable(vref), X`value is not durable`),
+        );
+      }
+      vrm.updateReferenceCounts(before.slots, after.slots);
+      syscall.vatstoreSet(dbKey, JSON.stringify(after));
+    }
+
+    function deleteInternal(key) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      const dbKey = keyToDBKey(key);
+      const rawValue = syscall.vatstoreGet(dbKey);
+      assert(rawValue, X`key ${key} not found in collection ${q(label)}`);
+      const value = JSON.parse(rawValue);
+      value.slots.map(vrm.removeReachableVref);
+      syscall.vatstoreDelete(dbKey);
+      let doMoreGC = false;
+      if (passStyleOf(key) === 'remotable') {
+        if (hasWeakKeys) {
+          vrm.removeRecognizableValue(key, entryDeleter);
+        } else {
+          doMoreGC = vrm.removeReachableVref(getSlotForVal(key));
+        }
+        deleteOrdinal(key);
+      }
+      ensureEntryCount();
+      entryCount -= 1;
+      return doMoreGC;
+    }
+
+    function del(key) {
+      currentGenerationNumber += 1;
+      deleteInternal(key);
+    }
+
+    function entriesInternal(
+      needValue,
+      keyPatt = M.any(),
+      valuePatt = M.any(),
+    ) {
+      assertKeyPattern(keyPatt);
+      assertPattern(valuePatt);
+      const [coverStart, coverEnd] = getRankCover(keyPatt, encodeKey);
+      let priorDBKey = '';
+      const start = prefix(coverStart);
+      const end = prefix(coverEnd);
+      const ignoreValues = !needValue && pattEq(valuePatt, M.any());
+      function* iter() {
+        const generationAtStart = currentGenerationNumber;
+        while (priorDBKey !== undefined) {
+          assert(generationAtStart === currentGenerationNumber);
+          const getAfterResult = syscall.vatstoreGetAfter(
+            priorDBKey,
+            start,
+            end,
+          );
+          if (!getAfterResult) {
```

Comment:
Hm, `getAfter` either returns `undefined` or `[key, value]`, which leads to some awkward unpacking conditionals. Should we consider (in a different PR, of course) making it return `[undefined, undefined]` for the end-of-iteration case? Would that make this scan better?

Reply:
Yes, I noticed this awkwardness as well when I was using it.  If we made the end condition `[undefined, undefined]` we'd still have to check the `key` but you could do the destructuring unconditionally and it would save adding another identifier to whatever scope we're in.  This would probably be cleaner, and the time to make this change is now before anybody else is really using this API.  I think I'll do that.